File: scripts/force_controller.py
import rospy
from geometry_msgs.msg import WrenchStamped, PoseStamped, TransformStamped
import message_filters
import tf2_ros

class ForceController:
    def __init__(self) -> None:
        
        self.Desired_force = 3 # 10 N desired force to applied
        self.Kp = 1
        self.Kd = 1
        self.Kf = 1000
        self.prev_error_fz = 0
        self.rate = rospy.Rate(10)
        self.dt = 1/10
        self.curr_pose = PoseStamped()
        self.error_fz = 0
        self.init_pose  = PoseStamped()
        self.init_pose.header.frame_id = "world"
        self.init_pose.header.stamp = rospy.Time.now()
        # self.init_pose.pose.position.x = -0.62
        # self.init_pose.pose.position.y = 0.0
        # self.init_pose.pose.position.z = 0.2
        # self.init_pose.pose.orientation.x = 0.7
        # self.init_pose.pose.orientation.y = 0.7
        # self.init_pose.pose.orientation.z = 0.0
        # self.init_pose.pose.orientation.w = 0.0
        buffer_ = tf2_ros.Buffer()
        listener = tf2_ros.TransformListener(buffer_)
        try:
            now = rospy.Time.now()
            transformation = buffer_.lookup_transform("iiwa_link_0", "tool_link_ee", rospy.Time(0), timeout=rospy.Duration(2))

        except (tf2_ros.LookupException, tf2_ros.ConnectivityException, tf2_ros.ExtrapolationException):
            print("exception occured")   

        self.init_pose.pose.position.x = transformation.transform.translation.x
        self.init_pose.pose.position.y = transformation.transform.translation.y
        self.init_pose.pose.position.z = transformation.transform.translation.z
        self.init_pose.pose.orientation.x = transformation.transform.rotation.x
        self.init_pose.pose.orientation.y = transformation.transform.rotation.y
        self.init_pose.pose.orientation.z = transformation.transform.rotation.z
        self.init_pose.pose.orientation.w = transformation.transform.rotation.w
        print(self.init_pose)

        self.pose_sub = message_filters.Subscriber("/tool_link_ee_pose", TransformStamped)    
        self.force_sub = message_filters.Subscriber("/cartesian_wrench_tool", WrenchStamped)
        # self.force_sub = message_filters.Subscriber("/ft_sensor/raw", WrenchStamped)
        # self.pose_pub = rospy.Publisher("/cartesian_trajectory_generator/new_goal", PoseStamped, queue_size=1)
        self.pose_pub = rospy.Publisher("/iiwa/CartesianImpedance_trajectory_controller/reference_pose", PoseStamped, queue_size=1)
        self.ts = message_filters.ApproximateTimeSynchronizer([self.pose_sub, self.force_sub], 1, 0.1, allow_headerless=True)
        self.ts.registerCallback(self.callback)


    # def current_pose(self, pose_msg):
    #     self.curr_pose.header.frame_id = "world"
    #     self.curr_pose.header.stamp = rospy.Time.now()
    #     self.curr_pose.pose.position.x = pose_msg.transform.translation.x
    #     self.curr_pose.pose.position.y = pose_msg.transform.translation.y
    #     self.curr_pose.pose.position.z = pose_msg.transform.translation.z
    #     self.curr_pose.pose.orientation.x = pose_msg.transform.rotation.x
    #     self.curr_pose.pose.orientation.y = pose_msg.transform.rotation.y
    #     self.curr_pose.pose.orientation.z = pose_msg.transform.rotation.z
    #     self.curr_pose.pose.orientation.w = pose_msg.transform.rotation.w



    def callback(self, pose_msg, force_msg):
        # self.rate.sleep()
        # self.curr_pose.header.frame_id = "world"
        # self.curr_pose.header.stamp = rospy.Time.now()
        # self.curr_pose.pose.position.x = pose_msg.transform.translation.x
        # self.curr_pose.pose.position.y = pose_msg.transform.translation.y
        # self.curr_pose.pose.position.z = pose_msg.transform.translation.z
        # self.curr_pose.pose.orientation.x = pose_msg.transform.rotation.x
        # self.curr_pose.pose.orientation.y = pose_msg.transform.rotation.y
        # self.curr_pose.pose.orientation.z = pose_msg.transform.rotation.z
        # self.curr_pose.pose.orientation.w = pose_msg.transform.rotation.w

        
        self.curr_force = force_msg.wrench.force.z

        print("current_force", self.curr_force)
        self.error_fz = self.Desired_force - abs(self.curr_force)
        print("self.error_fz", self.error_fz)
        dFe = self.error_fz - self.prev_error_fz
        # print("dFe", dFe)
        # print("prev_error", self.prev_error_fz)
        self.prev_error_fz = self.error_fz
        # dZ = (self.Kp*self.error_fz + self.Kd*dFe)*self.dt
        dZ = (self.Kp/self.Kf)*self.error_fz*self.dt + (self.Kd/self.Kf)*dFe*self.dt
        # print("dZ",dZ)
        self.update_pose(dZ)    
  
    
    def update_pose(self, dZ):
        # if (self.curr_force < 10):
        # print("poseinZ_before",self.init_pose.pose.position.z)
        self.init_pose.pose.position.z -= dZ
        # print("poseinZ",self.init_pose.pose.position.z)
        # print("dZ_again", dZ)
        # self.init_pose.pose.position.x = -0.62
        # self.init_pose.pose.position.y = 0.0
        # self.init_pose.pose.orientation.x = 0.7
        # self.init_pose.pose.orientation.y = 0.7
        # self.init_pose.pose.orientation.z = 0.0
        # self.init_pose.pose.orientation.w = 0.0
        
        # print(self.init_pose)
        self.pose_pub.publish(self.init_pose)
        # rospy.sleep(0.5)

        # if (self.curr_force > 10):
        #     self.init_pose.pose.position.z += dZ
        #     print(self.init_pose.pose.position.z)
        #     self.pose_pub.publish(self.init_pose)
            # rospy.sleep(0.5)
    
         

if __name__ == '__main__':
    rospy.init_node("force_controller", anonymous=True)

    ForceController()
    rospy.spin()

    

File: src/force_controller.cpp
#include <ros/ros.h>
#include <tf2_ros/static_transform_broadcaster.h>
#include <geometry_msgs/TransformStamped.h>
#include <cstdio>
#include <tf2/LinearMath/Quaternion.h>
#include "tf2_ros/transform_listener.h"
#include "tf2_geometry_msgs/tf2_geometry_msgs.h"
#include "message_filters/subscriber.h"
#include "geometry_msgs/WrenchStamped.h"
#include <cstdlib>
#include <tf2_ros/static_transform_broadcaster.h>
class ForceController{
    private:
    double  curr_force, error_fz, dFe, dZ;
    ros::Publisher pose_pub;
    ros::Subscriber wrench_sub;
    tf2::Transform transform_base_ee;
    geometry_msgs::TransformStamped transformStamped_base_to_end;
    geometry_msgs::TransformStamped transformStamped_goal;
    float desired_force=10.0;
    double Kp = 1.0;
    double Kd = 1.0;
    double Kf = 1000.0;
    double dt = 0.1;
    double prev_error_fz = 0.0;
    
    public:
    ForceController(ros::NodeHandle *nh){
      
        tf2_ros::Buffer tfBuffer;
        tf2_ros::TransformListener tfListener(tfBuffer);
        //looking up transformation from base to end-effector of the robot
        try {
                transformStamped_base_to_end = tfBuffer.lookupTransform("iiwa_link_0", "tool_link_ee", ros::Time(0), ros::Duration(2));
            }
        catch (tf2::TransformException &ex) {
                ROS_WARN("%s",ex.what());    
        }
        
        std::cout << transformStamped_base_to_end << std::endl;
        pose_pub = nh->advertise<geometry_msgs::PoseStamped>("/iiwa/CartesianImpedance_trajectory_controller/reference_pose",1);
        wrench_sub = nh->subscribe("/cartesian_wrench_tool",10, &ForceController::callback_controller, this);
}


    void callback_controller(const geometry_msgs::WrenchStamped& force_msg){
        
        curr_force = force_msg.wrench.force.z;
        std::cout << "current_force" << curr_force << std::endl;
        error_fz = desired_force - abs(curr_force);
         std::cout << "Fz error" << error_fz << std::endl;
        dFe = error_fz - prev_error_fz;
        prev_error_fz = error_fz;
        dZ = (Kp/Kf)*error_fz*dt + (Kd/Kf)*dFe*dt;        
        std::cout << "dZ" << dZ << std::endl;
        geometry_msgs::PoseStamped pose = update_pose(dZ);
        pose_pub.publish(pose);
        
    }


    geometry_msgs::PoseStamped update_pose(double& dZ){
        transformStamped_goal.header.stamp = ros::Time::now();
        transformStamped_goal.header.frame_id = "tool_link_ee";
        transformStamped_goal.child_frame_id = "goal_point";
        std::cout << "prev_z" << transformStamped_goal.transform.translation.z << std::endl;
        transformStamped_goal.transform.translation.x = 0.0;
        transformStamped_goal.transform.translation.y = 0.0;
        transformStamped_goal.transform.translation.z += dZ;
        transformStamped_goal.transform.rotation.x = 0.0;
        transformStamped_goal.transform.rotation.y = 0.0;
        transformStamped_goal.transform.rotation.z = 0.0;
        transformStamped_goal.transform.rotation.w = 1; 
        std::cout << transformStamped_goal << std::endl;
        tf2::Vector3 translation_goal(transformStamped_goal.transform.translation.x,
                                        transformStamped_goal.transform.translation.y,
                                        transformStamped_goal.transform.translation.z);

        tf2::Quaternion quat_tf_goal;
        tf2::convert(transformStamped_goal.transform.rotation, quat_tf_goal);
        quat_tf_goal.normalize();
        tf2::Transform transform_goal(quat_tf_goal,translation_goal);

        tf2::Vector3 translation_base_ee(transformStamped_base_to_end.transform.translation.x,transformStamped_base_to_end.transform.translation.y,transformStamped_base_to_end.transform.translation.z);
        tf2::Quaternion quat_tf_base_end;
        tf2::convert(transformStamped_base_to_end.transform.rotation, quat_tf_base_end);
        quat_tf_base_end.normalize();
        tf2::Transform transform_base_ee(quat_tf_base_end,translation_base_ee);

        tf2:: Transform transformed_goal_base = transform_base_ee*transform_goal;

        geometry_msgs::TransformStamped static_transform_goal_base;
        static_transform_goal_base.header.stamp = ros::Time::now();
        static_transform_goal_base.header.frame_id = "world";
        static_transform_goal_base.child_frame_id = "goal_frame";
        static_transform_goal_base.transform = tf2::toMsg(transformed_goal_base);

        geometry_msgs::PoseStamped pose_got;
        pose_got.header.frame_id="world";
        pose_got.header.stamp = ros::Time::now();
        pose_got.pose.position.x = static_transform_goal_base.transform.translation.x;
        pose_got.pose.position.y = static_transform_goal_base.transform.translation.y;
        pose_got.pose.position.z = static_transform_goal_base.transform.translation.z;
        pose_got.pose.orientation.x = static_transform_goal_base.transform.rotation.x;
        pose_got.pose.orientation.y = static_transform_goal_base.transform.rotation.y;
        pose_got.pose.orientation.z = static_transform_goal_base.transform.rotation.z;
        pose_got.pose.orientation.w = static_transform_goal_base.transform.rotation.w;
        // std::cout << pose_got << std::endl;
        return pose_got;
        


    }
};
        
int main(int argc, char **argv)
{
    ros::init(argc,argv, "force_controller");
    ros::NodeHandle nh;
    ForceController fc = ForceController(&nh);
    ros::spin();
}
